### PR TITLE
Bug fix 3.7/database initial syncer needs patience at times

### PR DESCRIPTION
**Pull request by kvahed**
### Scope & Purpose

*Wait for getting keys for initial sync more intelligently*

- [x] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [x] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [x] :book: CHANGELOG entry made

#### Backports:

- [ ] No backports required
- [ ] Backports required for: *(Please specify versions and link PRs)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket number:
- [ ] Design document: 

### Testing & Verification

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [ ] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [ ] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [x] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

Link to Jenkins PR run:
https://jenkins01.arangodb.biz/view/PR/job/arangodb-matrix-pr/13549/

### Documentation

> All new features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the CHANGELOG so that
> support, end users, and other developers have a concise overview. 

- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *HTTP API* 
- [ ] Added *Swagger examples* for the HTTP API  
- [ ] Updated license information in *LICENSES-OTHER-COMPONENTS.md* for 3rd party libraries

### External contributors / CLA Note 

Please note that for legal reasons we require you to sign the [Contributor Agreement](https://www.arangodb.com/documents/cla.pdf)
before we can accept your pull requests.

### CODE REVIEW

File: arangod/Replication/DatabaseInitialSyncer.cpp [hunk 1]
```diff
@@ -1078,104 +1078,138 @@ Result DatabaseInitialSyncer::fetchCollectionSyncByKeys(arangodb::LogicalCollect
     _config.barrier.extend(_config.connection);
   }
 
+  // We'll do one quick attempt at getting the keys on the leader.
+  // We might receive the keys or a count of the keys. In the first case we continue
+  // with the sync. If we get a coument count, we'll estimate a pessimistic wait and
```
Comment (neunhoef) [inline arangod/Replication/DatabaseInitialSyncer.cpp:1083]: typo in comment: coument -> document

File: arangod/Replication/DatabaseInitialSyncer.cpp [hunk 2]
```diff
@@ -1078,104 +1078,138 @@ Result DatabaseInitialSyncer::fetchCollectionSyncByKeys(arangodb::LogicalCollect
...
+    // side
+    // sending this request in a blocking fashion may require very long to
+    // complete,
+    // so we're sending the x-arango-async header here
+    auto headers = replutils::createHeaders();
+    headers[StaticStrings::Async] = "store";
 
-  bool found = false;
-  std::string jobId = response->getHeaderField(StaticStrings::AsyncId, found);
+    LOG_DEVEL << ((quick) ? url + "&quick=true" : url);
```
Comment (neunhoef) [inline arangod/Replication/DatabaseInitialSyncer.cpp:1109]: Either remove the LOG_DEVEL or change it into a debug log.

**Review by neunhoef — CHANGES_REQUESTED** [review]
LGTM, but see two small comments.

File: arangod/Replication/DatabaseInitialSyncer.cpp (hunk 2 above)
Comment (kvahed) [reply arangod/Replication/DatabaseInitialSyncer.cpp:1109]: Surely

**Review by neunhoef — APPROVED** [review]
LGTM

File: arangod/Replication/DatabaseInitialSyncer.cpp [hunk 3]
```diff
@@ -1078,104 +1078,136 @@ Result DatabaseInitialSyncer::fetchCollectionSyncByKeys(arangodb::LogicalCollect
...
+  };
+
+  auto ck = keysCall(true);
+  if (!ck.ok()) {
+    return ck;
+  }
+  if (!slice.hasKey("id")) { // we only have count
+    VPackSlice const c = slice.get("count");
+    if (c.isNumber()) {
+      maxWaitTime = c.getNumber<uint64_t>() * 8 / 100000;
```
Comment (jsteemann) [inline arangod/Replication/DatabaseInitialSyncer.cpp:1201]: IMHO `maxWaitTime` should be in microseconds, not in seconds.
If `count` contains a value less than 12.5K, the wait time will always be 0 here.
I think it should also have a minimum lower bound, e.g. `maxWaitTime = std::min(maxWaitTime, calculatedValue)`.

File: arangod/RocksDBEngine/RocksDBRestReplicationHandler.cpp [hunk 4]
```diff
@@ -402,15 +402,25 @@ void RocksDBRestReplicationHandler::handleCommandInventory() {
...
   if (collection.empty()) {
     generateError(rest::ResponseCode::BAD, TRI_ERROR_HTTP_BAD_PARAMETER,
                   "invalid collection parameter");
     return;
   }
+
+  std::string const& quick = _request->value("quick");
+  if (!quick.empty() && !(quick == "true" || quick == "false")) {
+    generateError(rest::ResponseCode::BAD, TRI_ERROR_HTTP_BAD_PARAMETER,
+                  std::string("invalid quick parameter: must be booleaan got ") + quick);
```
Comment (jsteemann) [inline arangod/RocksDBEngine/RocksDBRestReplicationHandler.cpp:419]:
```suggestion
                  std::string("invalid quick parameter: must be boolean, got ") + quick);
```

File: CHANGELOG [hunk 5]
```diff
@@ -21,14 +21,16 @@ v3.7.7 (XXXX-XX-XX)
 v3.7.6 (2021-01-04)
 -------------------
 
+* Fix too early stop of replication, when waiting for keys
```
Comment (jsteemann) [inline CHANGELOG:24]:
```suggestion
* Fix too early stop of replication, when waiting for keys in large collections/shards.
```

**Review by jsteemann — CHANGES_REQUESTED** [review]
We should also add tests for the new case.

**Review by neunhoef — APPROVED** [review]
LGTM

**Review by neunhoef — APPROVED** [review]
LGTM

**Comment by KVS85** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/14025/

**Comment by KVS85** [comment]
Tests blue except unrelated known failure.

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/14040/

**Comment by KVS85** [comment]
Tests blue.

**Review by jsteemann — APPROVED** [review]

**Pull request merged**